# Overview
## Background
The real estate industry in King County, Washington, plays a pivotal role in the region's economic landscape, contributing significantly to its growth and development. The housing market in this area is influenced by various factors such as location desirability, property size, amenities, and prevailing market trends. King County is renowned for its robust real estate market, offering diverse housing options ranging from single-family homes to condominiums and townhouses. The county has historically experienced strong demand due to a thriving economy, a robust job market, and the presence of major tech companies like Microsoft and Amazon. Its scenic beauty, including picturesque landscapes, waterfront properties, and a blend of urban and suburban neighborhoods, further adds to its appeal. King County also provides an array of amenities, including parks, recreational facilities, and cultural attractions, making it an attractive location for both homebuyers and investors.

## Business Understanding
The real estate industry in King County, Washington, plays a pivotal role in the region's economic landscape, contributing significantly to its growth and development. The housing market in this area is influenced by various factors such as location desirability, property size, amenities, and prevailing market trends. King County is renowned for its robust real estate market, offering diverse housing options ranging from single-family homes to condominiums and townhouses. The county has historically experienced strong demand due to a thriving economy, a robust job market, and the presence of major tech companies like Microsoft and Amazon. Its scenic beauty, including picturesque landscapes, waterfront properties, and a blend of urban and suburban neighborhoods, further adds to its appeal. King County also provides an array of amenities, including parks, recreational facilities, and cultural attractions, making it an attractive location for both homebuyers and investors.

## Challenges

## Solution
- Data Preparation. We will start by preparing the dataset, handling missing values, and encoding categorical variables to ensure that our data is ready for analysis. 
- Exploratory Data Analysis. We will conduct thorough exploratory data analysis to understand data variability, identify outliers, assess heteroscedasticity and detect multicollinearity. EDA will provide essential insights into the dataset and prepare us for subsequent regression analysis. 
- Regression Analysis. Building a multiple linear regression model will be our next step. We will select relevant features identified during EDA, address multicollinearity, and assess the model's performance to ensure that it meets the assumptions of linear regression

## Conclusion
This project focuses on addressing key challenges in the analysis of the King County real estate market. We aim to provide actionable recommendations to homeowners regarding the impact of home renovations on their property's estimated value. By matching our objectives to these challenges, we ensure a well-structured approach to data understanding, EDA, and regression analysis.The results of this analysis will not only benefit homeowners but also provide valuable guidance to real estate professionals in King County, helping them navigate this dynamic and diverse real estate market with confidence.

# Problem Statement
The challenge is to establish a robust data analysis framework that addresses the complexities including data preparation, exploratory data analysis and regression modeling. Our aim is to offer or provide homeowners and real estate professionals with dependable insights for navigating the ever-changing King County real estate market with confidence.

# Data Understanding
This project uses King County house sales dataset which can be found in kc_house_data.csv. 

This data contains information about the houses in the northwestern county. 

It includes the below features:
- price 
- bedrooms
- zipcode
- year built

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import scipy.stats as stats
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt
plt.style.use('seaborn')
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [2]:
# Load the dataset
df_housing = pd.read_csv("data/kc_house_data (1).csv")
df_housing

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0          5650     1.0        NaN  NONE  ...      7 Average       1180   
1          7242     2.0         NO  NONE  ...      7 Average       2170   
2         10000     1.0         NO  NONE  ...  6 Low Average        770   
3          5000     1.0         NO  NONE  ...      7 Average       1050   
4          8080     1.0         NO  NONE  ...         8 Good       1680   
...         ...     ...        ...   ...  ...            ...        ...   
21592      1131     3.0         NO  NONE  ...         8 Good       1530   
21593      5813     2.0         NO  NONE  ...         8 Good       2310   
21594      1350     2.0         NO  NONE  ...      7 Average       1020   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   
21596      1076     2.0         NO  NONE  ...      7 Average       1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0                0.0     1955           0.0    98178  47.5112 -122.257   
1              400.0     1951        1991.0    98125  47.7210 -122.319   
2                0.0     1933           NaN    98028  47.7379 -122.233   
3              910.0     1965           0.0    98136  47.5208 -122.393   
4                0.0     1987           0.0    98074  47.6168 -122.045   
...              ...      ...           ...      ...      ...      ...   
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [3]:
# Checking the shape of the dataset
df_housing.shape

(21597, 21)

In [4]:
# Obtain summary of the descriptive statistics for each numerical column 
df_housing.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

The function provides descriptive statistics for the numerical columns, including count, mean, standard deviation, minimum, maximum, and quartile values.

In [5]:
# Checking the dataframe structure and contents
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

The DataFrame contains 21597 rows and 21 columns.

The DataFrame consists of the columns 'id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long', 'sqft_living15', and 'sqft_lot15'.

The data types in the dataframe are floats, integers and objects.

The columns waterfront, view and yr_renovated contain missing values.

# Data Cleaning
To clean the data in preparation for analysis, we start with :

1. Check duplicates in the 'id' column.
2. Drop duplicates if necessary.
3. Identify and handle NAN (Not a Number) or missing values.
4. Check for place holders in 'price'column i.e 0.00
5. Convert date data types if necessary.
6. Identify outliers and either drop or keep them depending on the study objective.
7. Determining columns that are irrelevant for the analysis and drop them.

In [6]:
# Check for duplicates in the id column
duplicate_id = df_housing.duplicated(subset = "id")
print(f"""
We have a total of {duplicate_id.sum()} duplicates out of {df_housing.shape[0]} entries.
""")


We have a total of 177 duplicates out of 21597 entries.



We drop the entries with duplicated IDs to ensure our dataset accurately represents the unique houses and to avoid redundancy that might skew results which can lead to inaccurate analysis.

In [7]:
# Drop the duplicate rows
df_housing.drop_duplicates(subset = "id", inplace = True, keep = "first")
df_housing

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0          5650     1.0        NaN  NONE  ...      7 Average       1180   
1          7242     2.0         NO  NONE  ...      7 Average       2170   
2         10000     1.0         NO  NONE  ...  6 Low Average        770   
3          5000     1.0         NO  NONE  ...      7 Average       1050   
4          8080     1.0         NO  NONE  ...         8 Good       1680   
...         ...     ...        ...   ...  ...            ...        ...   
21592      1131     3.0         NO  NONE  ...         8 Good       1530   
21593      5813     2.0         NO  NONE  ...         8 Good       2310   
21594      1350     2.0         NO  NONE  ...      7 Average       1020   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   
21596      1076     2.0         NO  NONE  ...      7 Average       1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0                0.0     1955           0.0    98178  47.5112 -122.257   
1              400.0     1951        1991.0    98125  47.7210 -122.319   
2                0.0     1933           NaN    98028  47.7379 -122.233   
3              910.0     1965           0.0    98136  47.5208 -122.393   
4                0.0     1987           0.0    98074  47.6168 -122.045   
...              ...      ...           ...      ...      ...      ...   
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21420 rows x 21 columns]

In [8]:
# Identify missing values
df_housing.isnull().sum()

id                  0
date                0
price               0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       2353
view               63
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     3804
zipcode             0
lat                 0
long                0
sqft_living15       0
sqft_lot15          0
dtype: int64

In [9]:
# Calculating the percentage of missing values in the 'yr_renovated' column
df_housing["yr_renovated"].isna().mean() * 100

17.759103641456583

The missing values likely indicate that those particular houses have never been renovated. Therefore, we fill the null values with zeros to accurately reflect the absence of any renovation for those properties. This replacement ensures that the data remains consistent and that any subsequent analysis or modeling accurately accounts for the lack of renovation for those specific houses.

In [10]:
# Fill missing values with 0
df_housing["yr_renovated"].fillna(0, inplace = True)

In [11]:
# Checking if the null values have been filled
df_housing["yr_renovated"].isna().sum()

0

We now check for possible placeholders in the "price" column.

In [12]:
# Check unique values in the price column
unique_prices = df_housing['price'].unique()

# Sort the unique values in ascending order
sorted_prices = sorted(unique_prices)
sorted_prices[0]

78000.0

In [13]:
df_housing['price'].describe()

count    2.142000e+04
mean     5.407393e+05
std      3.679311e+05
min      7.800000e+04
25%      3.225000e+05
50%      4.500000e+05
75%      6.450000e+05
max      7.700000e+06
Name: price, dtype: float64

No place holders were identified in the price column.

The minimum price is 78,000. The maximum price is 7,700,000 . 

This figure sounds more as an outlier considering the distribution of the data and mean figures.

In [14]:
# Convert the date data type to datetime
df_housing["date"] = pd.to_datetime(df_housing["date"])
print(df_housing["date"].dtype)

datetime64[ns]


In [15]:
#create a function to find outliers using IQR
def find_outliers_IQR(df):
    q1=df.quantile(0.25)
    q3=df.quantile(0.75)
    IQR=q3-q1
    outliers = df[((df<(q1-1.5*IQR)) | (df>(q3+1.5*IQR)))]
    return outliers

In [16]:
outliers = find_outliers_IQR(df_housing["price"])

print("number of outliers: "+ str(len(outliers)))

print("max outlier value: "+ str(outliers.max()))

print("min outlier value: "+ str(outliers.min()))

outliers

number of outliers: 1151
max outlier value: 7700000.0
min outlier value: 1130000.0


5        1230000.0
21       2000000.0
49       1350000.0
69       1330000.0
125      1450000.0
           ...    
21552    1700000.0
21560    3570000.0
21574    1220000.0
21581    1580000.0
21584    1540000.0
Name: price, Length: 1151, dtype: float64

In [17]:
# Obtaining the count of each unique value in the "condition" column
df_housing["condition"].value_counts()

Average      13900
Good          5643
Very Good     1687
Fair           162
Poor            28
Name: condition, dtype: int64

Since the condition column is categorical, we convert the values to corresponding numerical representations by mapping the string values  'Poor', 'Fair', 'Average', 'Good', and 'Very Good'  to the values 1, 2, 3, 4, and 5.

In [18]:
# Map the categorical "condition" values to numerical
df_housing['condition'] = df_housing['condition'].replace('Very Good', 'Very_Good') # or 'VeryGood'
df_housing.loc[:,'condition'] = df_housing['condition'].map({'Poor': 1, 'Fair': 2, 'Average': 3, 'Good': 4, 'Very_Good': 5})
print(df_housing['condition'].unique())

[3 5 4 1 2]


We remove the non-numerical characters from the "grade" column.

In [19]:
# Remove non-numeric characters from the "grade" column
df_housing["grade"] = df_housing["grade"].str.replace("\D", "", regex = True)
df_housing["grade"] = df_housing["grade"].astype(int)

In [20]:
# Check that it has worked
df_housing["grade"].unique()

array([ 7,  6,  8, 11,  9,  5, 10, 12,  4,  3, 13])

We now drop the columns that we do not require for our analysis and check our final dataframe.

In [21]:
# Drop the columns that we do not require 
to_drop = ["sqft_living15", "lat", "long", "sqft_lot15", "zipcode", "view", "waterfront"]
df_housing.drop(to_drop, axis = 1, inplace = True)
df_housing 

id       date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520 2014-10-13  221900.0         3       1.00         1180   
1      6414100192 2014-12-09  538000.0         3       2.25         2570   
2      5631500400 2015-02-25  180000.0         2       1.00          770   
3      2487200875 2014-12-09  604000.0         4       3.00         1960   
4      1954400510 2015-02-18  510000.0         3       2.00         1680   
...           ...        ...       ...       ...        ...          ...   
21592   263000018 2014-05-21  360000.0         3       2.50         1530   
21593  6600060120 2015-02-23  400000.0         4       2.50         2310   
21594  1523300141 2014-06-23  402101.0         2       0.75         1020   
21595   291310100 2015-01-16  400000.0         3       2.50         1600   
21596  1523300157 2014-10-15  325000.0         2       0.75         1020   

       sqft_lot  floors  condition  grade  sqft_above sqft_basement  yr_built  \
0          5650     1.0          3      7        1180           0.0      1955   
1          7242     2.0          3      7        2170         400.0      1951   
2         10000     1.0          3      6         770           0.0      1933   
3          5000     1.0          5      7        1050         910.0      1965   
4          8080     1.0          3      8        1680           0.0      1987   
...         ...     ...        ...    ...         ...           ...       ...   
21592      1131     3.0          3      8        1530           0.0      2009   
21593      5813     2.0          3      8        2310           0.0      2014   
21594      1350     2.0          3      7        1020           0.0      2009   
21595      2388     2.0          3      8        1600           0.0      2004   
21596      1076     2.0          3      7        1020           0.0      2008   

       yr_renovated  
0               0.0  
1            1991.0  
2               0.0  
3               0.0  
4               0.0  
...             ...  
21592           0.0  
21593           0.0  
21594           0.0  
21595           0.0  
21596           0.0  

[21420 rows x 14 columns]

# Data Analysis